loopd/daemon.go

```diff
@@ -119,117 +135,291 @@ func daemon(config *config, lisCfg *listenerCfg) error {
+	swapsList, err := d.impl.FetchSwaps()
+	if err != nil {
+		return err
+	}
+
+	for _, s := range swapsList {
+		d.swaps[s.SwapHash] = *s
+	}
+
+	// Start the main event handlers, now that we have prepared everything
+	// they need to work.
+	d.startSwapHandlers()
+
+	return nil
+}
+
+// startSwapHandlers starts all goroutines that are tasked with handling swap
+// events.
+func (d *Daemon) startSwapHandlers() {
 	// Start the swap client itself.
-	wg.Add(1)
+	d.wg.Add(1)
 	go func() {
-		defer wg.Done()
-
 		log.Infof("Starting swap client")
-		err := swapClient.Run(mainCtx, server.statusChan)
+		err := d.impl.Run(d.mainCtx, d.statusChan)
 		if err != nil {
+			// Only log the error here for now. We'll send it over
+			// the errChan later, after we've stopped the daemon.
 			log.Error(err)
 		}
 		log.Infof("Swap client stopped")
 
-		log.Infof("Stopping gRPC server")
-		grpcServer.Stop()
+		// We need to mark this goroutine to be done explicitly here.
+		// Otherwise the stop below will wait for the defer which will
+		// end in a deadlock.
+		d.wg.Done()
+
+		// In case we got here because the swap client crashed, we need
+		// to shutdown as cleanly as possible. If we got here because of
+		// a normal shutdown, stop will prevent itself from running
+		// twice.
+		stopErr := d.Stop()
+		if stopErr != nil {
+			log.Error(stopErr)
+		}
+
+		switch {
+		// If we did crash for any reason, we want to return that error.
+		case err != nil:
+			d.errChan <- err
 
-		cancel()
+		// Normal shutdown, send any errors happening on stopping.
+		case stopErr != nil:
+			d.errChan <- stopErr
+
+		// No error happened, just close the error channel instead.
+		default:
+			close(d.errChan)
+		}
 	}()
 
 	// Start a goroutine that broadcasts swap updates to clients.
-	wg.Add(1)
+	d.wg.Add(1)
 	go func() {
-		defer wg.Done()
+		defer d.wg.Done()
 
 		log.Infof("Waiting for updates")
-		server.processStatusUpdates(mainCtx)
+		d.processStatusUpdates(d.mainCtx)
 	}()
+}
 
-	// Start the grpc server.
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-
-		log.Infof("RPC server listening on %s", grpcListener.Addr())
+// Stop tries to gracefully shut down the daemon.
+func (d *Daemon) Stop() error {
+	// If the swap client crashes, stop is called to make sure everything
+	// is cleaned up. But this also means we get here twice in case of the
+	// normal shutdown because after the swap client exists normally, stop
+	// is called again. We don't want to log an error in that case.
+	if atomic.AddInt32(&d.stopped, 1) != 1 {
+		return nil
+	}
 
-		if restListener != nil {
-			log.Infof("REST proxy listening on %s", restListener.Addr())
-		}
+	// First of all, tell the gRPC server to stop accepting connections.
+	// This method blocks until all pending requests are completed. Because
+	// of our streaming subscriptions, those won't be canceled until the
+	// main context is canceled, so we'd be stuck here. That's why we need
+	// to run this in a goroutine.
```

Comment:
If this is the case anyway, it looks indeed to be better to first cancel the main context and not start this goroutine. Maybe this only makes sense if there would be a delay before cancelling the main context. But that may be seen as annoying if there are not pending grpc connections. Maybe it should give `GracefulStop` up to a specific time to stop gracefully and thereafter we hard-cancel the context.

Reply:
But the worry was that while the main context is being canceled and the swap handlers are shutting down, we would still accept incoming requests. This is handled by this now.
Or do you think we should add an explicit delay after the goroutine with the graceful shutdown was started?

Reply:
Is it really handled? To me it looks like we call `GracefulStop` and at the exact same moment we cancel the main context. Which basically means that the order isn't really changed from before? There may also still a request "in flight", but don't know how that works internally exactly.

To me the cleanest approach is what I described above. Give GracefulStop *up to* a specific time to stop gracefully and thereafter we hard-cancel the context.

But we can also decide not to deal with this at all and keep it simple without goroutines. First  cancel context, then stop grpc.

Reply:
Okay, removed the goroutine.